shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java

```diff
@@ -78,8 +87,16 @@ public void detachFromRenderSurface() {
     }
   }
 
+  public boolean hasRenderedFirstFrame() {
+    return hasRenderedFirstFrame;
+  }
+
   public void addOnFirstFrameRenderedListener(@NonNull OnFirstFrameRenderedListener listener) {
     flutterJNI.addOnFirstFrameRenderedListener(listener);
```

Comment:
what does this do in this case? If a first frame already rendered with the renderer and another view calls this, it gets called back instantly and the listener gets registered through jni as a listener again? Does that listener get called again at some point?

Reply:
I think your description is correct. The listener when added to a `FlutterRenderer` should not be invoked again because a Dart isolate can't be entered twice. But that's kind of an implementation detail, and this listener interface is also used at the FlutterView level, which could be called many times. So generally speaking wherever such a listener is involved, we accept adding and removing of listeners, and we invoke them whenever is appropriate for that place.

However, there is some cleanup I'd like to do WRT to these first frame callbacks.